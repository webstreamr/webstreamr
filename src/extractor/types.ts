import { Context, UrlResult } from '../types';

export interface Extractor {
  readonly id: string;

  readonly label: string;

  readonly ttl: number;

  readonly supports: (url: URL) => boolean;

  readonly extract: (ctx: Context, url: URL, countryCode: string) => Promise<UrlResult>;
}
