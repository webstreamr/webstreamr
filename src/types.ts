import { Manifest, ManifestConfig } from 'stremio-addon-sdk';

export const TIMEOUT = 'timeout';

export interface Context {
  id: string;
  ip: string;
  config: Config;
  referer?: URL;
}

export type ManifestWithConfig = Manifest & { config: ManifestConfig[] };

export type Config = Partial<Record<CountryCode | 'excludeExternalUrls', string>>;

export type CountryCode = 'de' | 'en' | 'es' | 'fr' | 'it' | 'mx';

export type BlockedReason = 'cloudflare_challenge' | 'unknown';

export interface Meta {
  bytes?: number;
  countryCode: CountryCode;
  height?: number;
  title?: string;
}

export interface UrlResult {
  url: URL;
  isExternal?: boolean;
  error?: unknown;
  label: string;
  sourceId: string;
  ttl?: number;
  meta: Meta;
  requestHeaders?: Record<string, string>;
}
