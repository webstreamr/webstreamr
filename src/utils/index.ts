export * from './embed';
export * from './fetch';
export * from './imdb';
export * from './language';
export * from './log';
export * from './manifest';
export * from './promise';
