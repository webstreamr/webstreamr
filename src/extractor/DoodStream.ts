import randomstring from 'randomstring';
import * as cheerio from 'cheerio';
import { Extractor } from './types';
import { Fetcher } from '../utils';
import { Context, Meta } from '../types';
import { NotFoundError } from '../error';

export class DoodStream implements Extractor {
  readonly id = 'doodstream';

  readonly label = 'DoodStream';

  readonly ttl = 900000; // 15m

  private readonly fetcher: Fetcher;

  constructor(fetcher: Fetcher) {
    this.fetcher = fetcher;
  }

  readonly supports = (_ctx: Context, url: URL): boolean => null !== url.host.match(/dood|do[0-9]go/);

  readonly extract = async (ctx: Context, url: URL, meta: Meta) => {
    const videoId = url.pathname.split('/').slice(-1)[0] as string;
    const normalizedUrl = new URL(`http://dood.to/e/${videoId}`);

    const html = await this.fetcher.text(ctx, new URL(normalizedUrl));

    const passMd5Match = html.match(/\/pass_md5\/[\w-]+\/([\w-]+)/);
    if (!passMd5Match) {
      throw new NotFoundError();
    }

    const token = passMd5Match[1] as string;

    const baseUrl = await this.fetcher.text(ctx, new URL(`http://dood.to${passMd5Match[0]}`));

    const $ = cheerio.load(html);
    const title = $('title').text().trim().replace(/ - DoodStream$/, '').trim();

    return {
      url: new URL(`${baseUrl}${randomstring.generate(10)}?token=${token}&expiry=${Date.now()}`),
      label: this.label,
      sourceId: `${this.id}_${meta.countryCode.toLowerCase()}`,
      ttl: this.ttl,
      meta: {
        title,
        ...meta,
      },
      requestHeaders: {
        Referer: 'http://dood.to/',
      },
    };
  };
}
