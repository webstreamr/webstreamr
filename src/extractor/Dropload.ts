import bytes from 'bytes';
import * as cheerio from 'cheerio';
import { Extractor } from './types';
import { extractUrlFromPacked, Fetcher } from '../utils';
import { Context, Meta } from '../types';
import { NotFoundError } from '../error';

export class Dropload implements Extractor {
  readonly id = 'dropload';

  readonly label = 'Dropload';

  readonly ttl = 900000; // 15m

  private readonly fetcher: Fetcher;

  constructor(fetcher: Fetcher) {
    this.fetcher = fetcher;
  }

  readonly supports = (_ctx: Context, url: URL): boolean => null !== url.host.match(/dropload/);

  readonly extract = async (ctx: Context, url: URL, meta: Meta) => {
    url.pathname = url.pathname.replace('/e/', '').replace('/embed-', '/');
    const html = await this.fetcher.text(ctx, url);

    if (html.includes('File Not Found')) {
      throw new NotFoundError();
    }

    const heightMatch = html.match(/\d{3,}x(\d{3,}),/) as string[];

    const sizeMatch = html.match(/([\d.]+ ?[GM]B)/) as string[];

    const $ = cheerio.load(html);
    const title = $('.videoplayer h1').text().trim();

    return {
      url: extractUrlFromPacked(html, [/sources:\[{file:"(.*?)"/]),
      label: this.label,
      sourceId: `${this.id}_${meta.countryCode.toLowerCase()}`,
      ttl: this.ttl,
      meta: {
        bytes: bytes.parse(sizeMatch[1] as string) as number,
        height: parseInt(heightMatch[1] as string) as number,
        title,
        ...meta,
      },
    };
  };
}
