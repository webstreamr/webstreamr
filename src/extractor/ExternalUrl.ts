import { Extractor } from './types';
import { Fetcher } from '../utils';
import { Context, Meta } from '../types';

export class ExternalUrl implements Extractor {
  readonly id = 'external';

  readonly label = 'External';

  readonly ttl = 3600000; // 1h

  private readonly fetcher: Fetcher;

  constructor(fetcher: Fetcher) {
    this.fetcher = fetcher;
  }

  readonly supports = (ctx: Context, url: URL): boolean => !('excludeExternalUrls' in ctx.config) && null !== url.host.match(/.*/);

  readonly extract = async (ctx: Context, url: URL, meta: Meta) => {
    // We only want to make sure that the URL is accessible
    await this.fetcher.head(ctx, url);

    return {
      url: url,
      isExternal: true,
      label: `${url.host}`,
      sourceId: `${this.id}_${meta.countryCode.toLowerCase()}`,
      ttl: this.ttl,
      meta,
    };
  };
}
