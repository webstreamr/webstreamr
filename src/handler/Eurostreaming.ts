import { ContentType } from 'stremio-addon-sdk';
import * as cheerio from 'cheerio';
import { Handler } from './types';
import { ImdbId, Fetcher, Id, getImdbId } from '../utils';
import { ExtractorRegistry } from '../extractor';
import { Context, CountryCode } from '../types';

export class Eurostreaming implements Handler {
  readonly id = 'eurostreaming';

  readonly label = 'Eurostreaming';

  readonly contentTypes: ContentType[] = ['series'];

  readonly countryCodes: CountryCode[] = ['it'];

  private readonly fetcher: Fetcher;
  private readonly extractorRegistry: ExtractorRegistry;

  constructor(fetcher: Fetcher, extractorRegistry: ExtractorRegistry) {
    this.fetcher = fetcher;
    this.extractorRegistry = extractorRegistry;
  }

  readonly handle = async (ctx: Context, _type: string, id: Id) => {
    const imdbId = await getImdbId(ctx, this.fetcher, id);

    const seriesPageUrl = await this.fetchSeriesPageUrl(ctx, imdbId);
    if (!seriesPageUrl) {
      return [];
    }

    const html = await this.fetcher.text(ctx, seriesPageUrl);

    const $ = cheerio.load(html);

    const mainDataLinkElements = $(`[data-num="${imdbId.season}x${imdbId.episode}"][data-link!="#"]`);
    const mirrorDataLinkElements = $(`[data-num="${imdbId.season}x${imdbId.episode}"]`)
      .siblings('.mirrors')
      .children('[data-link!="#"]');

    const title = $('meta[property="og:title"]').attr('content') as string;

    return Promise.all(mainDataLinkElements
      .add(mirrorDataLinkElements)
      .map((_i, el) => new URL(($(el).attr('data-link') as string).replace(/^(https:)?\/\//, 'https://')))
      .toArray()
      .filter(url => !url.host.match(/eurostreaming/))
      .map(url => this.extractorRegistry.handle({ ...ctx, referer: seriesPageUrl }, url, { countryCode: 'it', title: `${title.trim()} ${imdbId.season}x${imdbId.episode}` })),
    );
  };

  private fetchSeriesPageUrl = async (ctx: Context, imdbId: ImdbId): Promise<URL | undefined> => {
    const html = await this.fetcher.textPost(
      ctx,
      new URL('https://eurostreaming.my/index.php'),
      JSON.stringify({
        do: 'search',
        subaction: 'search',
        search_start: 0,
        full_search: 0,
        result_from: 1,
        story: imdbId.id,
      }),
      {
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded',
        },
      },
    );

    const $ = cheerio.load(html);

    const url = $('.post-content a[href]:first')
      .map((_i, el) => $(el).attr('href'))
      .get(0);

    return url !== undefined ? new URL(url) : url;
  };
}
