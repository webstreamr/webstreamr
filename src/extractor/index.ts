import { Fetcher } from '../utils';
import { DoodStream } from './DoodStream';
import { Dropload } from './Dropload';
import { ExternalUrl } from './ExternalUrl';
import { Extractor } from './Extractor';
import { Fastream } from './Fastream';
import { FileLions } from './FileLions';
import { HubCloud } from './HubCloud';
import { KinoGer } from './KinoGer';
import { Mixdrop } from './Mixdrop';
import { SaveFiles } from './SaveFiles';
import { Soaper } from './Soaper';
import { StreamEmbed } from './StreamEmbed';
import { Streamtape } from './Streamtape';
import { SuperVideo } from './SuperVideo';
import { Uqload } from './Uqload';
import { VidSrc } from './VidSrc';
import { VixSrc } from './VixSrc';
import { XPrime } from './XPrime';
import { YouTube } from './YouTube';

export * from './Extractor';
export * from './ExtractorRegistry';

export const createExtractors = (fetcher: Fetcher): Extractor[] => [
  new DoodStream(fetcher),
  new Dropload(fetcher),
  new Fastream(fetcher),
  new FileLions(fetcher),
  new HubCloud(fetcher),
  new KinoGer(fetcher),
  new Mixdrop(fetcher),
  new SaveFiles(fetcher),
  new Soaper(fetcher),
  new StreamEmbed(fetcher),
  new Streamtape(fetcher),
  new SuperVideo(fetcher),
  new Uqload(fetcher),
  new VidSrc(fetcher, ['in', 'net', 'xyz', 'io', 'vc']), // https://vidsrc.domains/
  new VixSrc(fetcher),
  new XPrime(fetcher),
  new YouTube(fetcher),
  new ExternalUrl(fetcher), // fallback extractor which must come last
];
