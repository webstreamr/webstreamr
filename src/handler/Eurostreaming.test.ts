import winston from 'winston';
import { Eurostreaming } from './Eurostreaming';
import { ExtractorRegistry } from '../extractor';
import { Fetcher, ImdbId } from '../utils';
import { Context } from '../types';
jest.mock('../utils/Fetcher');

const logger = winston.createLogger({ transports: [new winston.transports.Console({ level: 'nope' })] });
// @ts-expect-error No constructor args needed
const fetcher = new Fetcher();
const ctx: Context = { id: 'id', ip: '127.0.0.1', config: { it: 'on' } };

describe('Eurostreaming', () => {
  let handler: Eurostreaming;

  beforeEach(() => {
    handler = new Eurostreaming(fetcher, new ExtractorRegistry(logger, fetcher));
  });

  test('handles non-existent series gracefully', async () => {
    const streams = await handler.handle(ctx, 'series', new ImdbId('tt12345678', 1, 1));
    expect(streams).toHaveLength(0);
  });

  test('handle imdb black mirror s2e4', async () => {
    const streams = await handler.handle(ctx, 'series', new ImdbId('tt2085059', 2, 4));
    expect(streams).toMatchSnapshot();
  });
});
