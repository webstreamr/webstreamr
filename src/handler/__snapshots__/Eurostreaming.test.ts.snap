// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Eurostreaming handle imdb black mirror s2e4 1`] = `
[
  {
    "label": "SuperVideo",
    "meta": {
      "bytes": 875875532,
      "countryCode": "it",
      "height": 1080,
      "title": "Black Mirror 2x4",
    },
    "sourceId": "supervideo_it",
    "ttl": 900000,
    "url": "https://hfs309.serversicuro.cc/hls/dnzpdeoe5xg4a3gyvaqx5ojpswtxjd5a22rklgah7,khdhascykjrixs7huqq,obtfascykj3vna7rk5a,.urlset/master.m3u8",
  },
  {
    "label": "SuperVideo",
    "meta": {
      "bytes": 875875532,
      "countryCode": "it",
      "height": 1080,
      "title": "Black Mirror 2x4",
    },
    "sourceId": "supervideo_it",
    "ttl": 900000,
    "url": "https://hfs309.serversicuro.cc/hls/dnzpdeoe5xg4a3gyvaqx5ojpswtxjd5a22rklgah7,khdhascykjrixs7huqq,obtfascykj3vna7rk5a,.urlset/master.m3u8",
  },
  {
    "label": "Dropload",
    "meta": {
      "bytes": 875875532,
      "countryCode": "it",
      "height": 1080,
      "title": "Black Mirror 2x4",
    },
    "sourceId": "dropload_it",
    "ttl": 900000,
    "url": "https://srv24.dropload.io/hls2/01/00289/978xyyxi3ldq_h/master.m3u8?t=unq9qxjYC-zgcmoXiSQ6rkMu7qHrSpN_pWOc95IPYwk&s=1747217151&e=14400&f=1448116&i=0.0&sp=0&ii=130.61.236.204",
  },
]
`;
